awslabs/aws-lc
# Test serialization of server SSL(TLS 1.2 after handshake finished).

### Issues:
Addresses CryptoAlg-850

### Description of changes: 
This PR added below changes:
* moved the code of `CR-46206164` to GitHub.
  * `CR-46206164` was submitted by `Evgeny Potemkin` and includes the initial implementation of SSL connection transfer(`i2d/d2i_SSL` APIs). 
* converted the existing unit and runner tests(see **Testing** section) to test if the decoded SSL has any behavior change.

### Call-outs:
* This PR only encodes fields based on the current scope(TLS 1.2 SSL as server after handshake completes).
* Fuzz tests will be published in new PRs.

### Testing:

Existing eligible(supported by current scope) unit and runner tests are converted to ensure the transferred SSL does not have behavior change.

#### Converted runner tests
See `ssl/test/runner/ssl_transfer/test_case_names.txt` in this PR
Test categories:
1. Ensure `SSL getter APIs`(e.g. `SSL_get0_alpn_selected`) have no behavior change after SSL transferred(encoded and then decoded). e.g. test case: `ALPNServer-Async-TLS-TLS12`
2. Ensure the `|SSL_peek/read|` API returns correct value when the original SSL has buffered read data. e.g.
```
# Test case names(below needs suffix|-SSL_Transfer|):
Peek-Basic-Server
Peek-Shutdown-Server
Peek-ImplicitHandshake-Server

# SSL client(Golang impl) sends messages to SSL server(bssl impl).
# The messages size is 1 + 16384(maximum plaintext payload length).
# SSL server keeps calling SSL_peek/read with size 512 until all messages are read.
# - In each iteration(peek/read), the SSL transfer is performed.
# - The peeked data is compared before and after the transfer to ensure correctness of serialization.
# Below shows the size of transferred |pending_app_data| and |read_buffer|.
-------------------------------
ssl->s3->pending_app_data size 16384
ssl->s3->read_buffer buf_size 16412
-------------------------------
ssl->s3->pending_app_data size 15872
ssl->s3->read_buffer buf_size 16412
-------------------------------
ssl->s3->pending_app_data size 15360
ssl->s3->read_buffer buf_size 16412
-------------------------------
ssl->s3->pending_app_data size 14848
ssl->s3->read_buffer buf_size 16412
-------------------------------
...
-------------------------------
ssl->s3->pending_app_data size 1024
ssl->s3->read_buffer buf_size 16412
-------------------------------
ssl->s3->pending_app_data size 512
ssl->s3->read_buffer buf_size 16412
-------------------------------
ssl->s3->pending_app_data size 1
ssl->s3->read_buffer buf_size 29
```
3. Ensure the session resumption of transferred SSL still works. A lot of test cases cover the session resumption.
4. Ensure the `SSL_read` and `SSL_write` still work with different(see below) cipher suites.
```
# Test case names(below needs suffix|-SSL_Transfer|)
CipherNegotiation-0
CipherNegotiation-1
CipherNegotiation-2
CipherNegotiation-3
CipherNegotiation-4
CipherNegotiation-5
CipherNegotiation-6
CipherNegotiation-7
CipherNegotiation-8
CipherNegotiation-9

# Negotiated cipher(returned by |SSL_get_cipher_name|).
AES128-GCM-SHA256
AES128-SHA
AES256-GCM-SHA384
AES256-SHA
DES-CBC3-SHA
ECDHE-ECDSA-AES128-GCM-SHA256
ECDHE-ECDSA-AES128-SHA
ECDHE-ECDSA-AES256-GCM-SHA384
ECDHE-ECDSA-AES256-SHA
ECDHE-ECDSA-CHACHA20-POLY1305
ECDHE-PSK-AES128-CBC-SHA
ECDHE-PSK-AES256-CBC-SHA
ECDHE-PSK-CHACHA20-POLY1305
ECDHE-RSA-AES128-GCM-SHA256
ECDHE-RSA-AES128-SHA
ECDHE-RSA-AES256-GCM-SHA384
ECDHE-RSA-AES256-SHA
ECDHE-RSA-CHACHA20-POLY1305
PSK-AES128-CBC-SHA
PSK-AES256-CBC-SHA
```

#### Converted unit tests
```
# WithVersion
WithVersion/SSLVersionTest.SequenceNumber/TLS1_2_SSL_TRANSFER
WithVersion/SSLVersionTest.OneSidedShutdown/TLS1_2_SSL_TRANSFER
WithVersion/SSLVersionTest.WriteAfterShutdown/TLS1_2_SSL_TRANSFER
WithVersion/SSLVersionTest.WriteAfterReadSentFatalAlert/TLS1_2_SSL_TRANSFER
WithVersion/SSLVersionTest.GetPeerCertificate/TLS1_2_SSL_TRANSFER
WithVersion/SSLVersionTest.NoPeerCertificate/TLS1_2_SSL_TRANSFER
WithVersion/SSLVersionTest.RetainOnlySHA256OfCerts/TLS1_2_SSL_TRANSFER
WithVersion/SSLVersionTest.SessionIDContext/TLS1_2_SSL_TRANSFER
WithVersion/SSLVersionTest.SessionTimeout/TLS1_2_SSL_TRANSFER
WithVersion/SSLVersionTest.DefaultTicketKeyRotation/TLS1_2_SSL_TRANSFER
WithVersion/SSLVersionTest.SNICallback/TLS1_2_SSL_TRANSFER
WithVersion/SSLVersionTest.Version/TLS1_2_SSL_TRANSFER
WithVersion/SSLVersionTest.ALPNCipherAvailable/TLS1_2_SSL_TRANSFER
WithVersion/SSLVersionTest.SSLClearSessionResumption/TLS1_2_SSL_TRANSFER
WithVersion/SSLVersionTest.SSLClearFailsWithShedding/TLS1_2_SSL_TRANSFER
WithVersion/SSLVersionTest.AutoChain/TLS1_2_SSL_TRANSFER
WithVersion/SSLVersionTest.SSLWriteRetry/TLS1_2_SSL_TRANSFER
WithVersion/SSLVersionTest.RecordCallback/TLS1_2_SSL_TRANSFER
WithVersion/SSLVersionTest.GetServerName/TLS1_2_SSL_TRANSFER
WithVersion/SSLVersionTest.ClientSessionCacheMode/TLS1_2_SSL_TRANSFER
WithVersion/SSLVersionTest.SmallBuffer/TLS1_2_SSL_TRANSFER
WithVersion/SSLVersionTest.SessionVersion/TLS1_2_SSL_TRANSFER
WithVersion/SSLVersionTest.SSLPending/TLS1_2_SSL_TRANSFER
WithVersion/SSLVersionTest.FakeIDsForTickets/TLS1_2_SSL_TRANSFER
WithVersion/SSLVersionTest.SessionCacheThreads/TLS1_2_SSL_TRANSFER
WithVersion/SSLVersionTest.SessionTicketThreads/TLS1_2_SSL_TRANSFER
WithVersion/SSLVersionTest.SessionPropertiesThreads/TLS1_2_SSL_TRANSFER
WithVersion/SSLVersionTest.DoubleSSLError/TLS1_2_SSL_TRANSFER
WithVersion/SSLVersionTest.SameKeyResume/TLS1_2_SSL_TRANSFER
WithVersion/SSLVersionTest.DifferentKeyNoResume/TLS1_2_SSL_TRANSFER
WithVersion/SSLVersionTest.UnrelatedServerNoResume/TLS1_2_SSL_TRANSFER
WithVersion/SSLVersionTest.TicketSessionIDsMatch/TLS1_2_SSL_TRANSFER

# SSLTests
SSLTests/SSLTest.WriteWhileExplicitRenegotiate/NO_SSL_Transfer
SSLTests/SSLTest.WriteWhileExplicitRenegotiate/SSL_Transfer

# TicketAEADMethodTests
TicketAEADMethodTests/TicketAEADMethodTest.Resume/TLS12_*_SSLTransfer
```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
